pkg/tfgen/generate_csharp.go

```diff
@@ -0,0 +1,1312 @@
+		}
+	} else {
+		if _, err := os.Stat(dst); err != nil {
+			return "", err
+		}
+	}
+
+	// And then export the overlay's contents from the index.
+	return dst, nil
+}
+
+func csharpIdentifier(s string) string {
+	switch s {
+	case "abstract", "as", "base", "bool",
+		"break", "byte", "case", "catch",
+		"char", "checked", "class", "const",
+		"continue", "decimal", "default", "delegate",
+		"do", "double", "else", "enum",
+		"event", "explicit", "extern", "false",
+		"finally", "fixed", "float", "for",
+		"foreach", "goto", "if", "implicit",
+		"in", "int", "interface", "internal",
+		"is", "lock", "long", "namespace",
+		"new", "null", "object", "operator",
+		"out", "override", "params", "private",
+		"protected", "public", "readonly", "ref",
+		"return", "sbyte", "sealed", "short",
+		"sizeof", "stackalloc", "static", "string",
+		"struct", "switch", "this", "throw",
+		"true", "try", "typeof", "uint",
+		"ulong", "unchecked", "unsafe", "ushort",
+		"using", "virtual", "void", "volatile", "while":
+		return "@" + s
+
+	default:
+		return s
+	}
+}
+
+func csharpPropertyName(v *variable) string {
+	if v.info != nil && v.info.CSharpName != "" {
+		return v.info.CSharpName
+	}
+	return strings.Title(v.name)
+}
+
+// csharpPropertyType returns the C# type name for a given schema property.
+//
+// wrapInput can be set to true to cause the generated type to be deeply wrapped with `Input<T>`.
+func csharpPropertyType(v *variable, wrapInput bool) string {
+	return qualifiedCSharpPropertyType(v, "", wrapInput)
+}
+
+func isImmutableArrayType(typ *propertyType, wrapInput bool) bool {
+	return (typ.kind == kindSet || typ.kind == kindList) && !wrapInput
+}
+
+func qualifiedCSharpPropertyType(v *variable, qualifier string, wrapInput bool) string {
+	t := qualifiedCSharpType(v.typ, qualifier, wrapInput)
+	if v.optional() && !isImmutableArrayType(v.typ, wrapInput) {
+		t += "?"
+	}
+	return t
+}
+
+// qualifiedCsharpType returns the C# type name for a given schema value type and element kind.
+func qualifiedCSharpType(typ *propertyType, qualifier string, wrapInput bool) string {
+	// Prefer overrides over the underlying type.
+	var t string
+	switch {
+	case typ == nil:
+		return "object"
+	case typ.asset != nil:
+		if typ.asset.IsArchive() {
+			t = typ.asset.Type()
+		} else {
+			t = "AssetOrArchive"
+		}
+	default:
+		// First figure out the raw type.
+		switch typ.kind {
+		case kindBool:
+			t = "bool"
+		case kindInt:
+			t = "int"
+		case kindFloat:
+			t = "double"
+		case kindString:
+			t = "string"
+		case kindSet, kindList:
+			elemType := qualifiedCSharpType(typ.element, qualifier, false)
+			if wrapInput {
+				return fmt.Sprintf("InputList<%v>", elemType)
+			}
+			return fmt.Sprintf("ImmutableArray<%v>", elemType)
+		case kindMap:
+			elemType := qualifiedCSharpType(typ.element, qualifier, false)
+			if wrapInput {
+				return fmt.Sprintf("InputMap<%v>", elemType)
+			}
+			return fmt.Sprintf("ImmutableDictionary<string, %v>", elemType)
```

Comment:
open thought: we shoudl probably rename the core types to `InputArray` and `InputDictionary` to be more consistent.